Zotero/Controllers/Attachment Downloader/AttachmentDownloader.swift

```diff
@@ -134,218 +343,551 @@ final class AttachmentDownloader {
+        if let response = downloadTask.response as? HTTPURLResponse {
+            let _isCompressed = response.value(forHTTPHeaderField: "Zotero-File-Compressed") == "Yes" || response.value(forHTTPHeaderField: "Content-Type") == "application/zip"
+            isCompressed = isCompressed || _isCompressed
+        }
+        if isCompressed {
+            zipFile = activeDownload.file.copyWithExt("zip")
+        } else {
+            shouldExtractAfterDownload = false
+        }
+
+        do {
+            if let error = checkFileResponse(for: Files.file(from: location)) {
+                throw error
+            }
+            // If there is some older version of given file, remove so that it can be replaced
+            if let zipFile, fileStorage.has(zipFile) {
+                try fileStorage.remove(zipFile)
+            }
+            if fileStorage.has(activeDownload.file) {
+                try fileStorage.remove(activeDownload.file)
+            }
+            // Move downloaded file to new location
+            try fileStorage.move(from: location.path, to: zipFile ?? activeDownload.file)
+
+            dbQueue.sync { [weak self] in
+                guard let self else { return }
+                // Mark file as downloaded in DB
+                let request = MarkFileAsDownloadedDbRequest(key: download.key, libraryId: download.libraryId, downloaded: true, compressed: isCompressed)
+                try? dbStorage.perform(request: request, on: dbQueue)
+            }
 
-            self.errors[download] = nil
-            self.observable.on(.next(Update(download: download, parentKey: parentKey, kind: .ready)))
+            accessQueue.sync(flags: .barrier) { [weak self] in
+                self?.finish(activeDownload: activeDownload, download: download, result: .success(!shouldExtractAfterDownload))
+            }
 
-        case .failure(let error):
-            DDLogError("AttachmentDownloader: failed to download attachment \(download.key), \(download.libraryId) - \(error)")
+            if let zipFile, shouldExtractAfterDownload {
+                extract(zipFile: zipFile, toFile: activeDownload.file, download: download)
+            }
+        } catch let error {
+            accessQueue.sync(flags: .barrier) { [weak self] in
+                self?.finish(activeDownload: activeDownload, download: download, result: .failure(error))
+            }
+        }
 
-            let isCancelError = (error as? AttachmentDownloadOperation.Error) == .cancelled
-            self.errors[download] = (isCancelError || hasLocalCopy) ? nil : error
+        func checkFileResponse(for file: File) -> Swift.Error? {
+            let size = self.fileStorage.size(of: file)
+            if size == 0 || (size == 9 && (try? self.fileStorage.read(file)).flatMap({ String(data: $0, encoding: .utf8) }) == "Not found") {
+                try? self.fileStorage.remove(file)
```

Comment:
```suggestion
            let size = fileStorage.size(of: file)
            if size == 0 || (size == 9 && (try? fileStorage.read(file)).flatMap({ String(data: $0, encoding: .utf8) }) == "Not found") {
                try? fileStorage.remove(file)
```
or, if you want to make it safe in case this is ever called asynchronously, make `filesStorage: FileStorage` a parameter